### PR TITLE
WIP: Initial README for new Example Notebook Server Dockerfiles and images

### PR DESCRIPTION
**Pull request by DavidSpek**
This is the README to go along with the new example Notebook Server Dockerfiles and images from https://github.com/kubeflow/kubeflow/issues/5575. The chart does not yet contain all the images from https://github.com/kubeflow/kubeflow/issues/5575 (it only contains those for with the PR has already been created), and could probably be made a bit prettier. I believe I've covered the important points that apply to all of the images new images. Where necessary I will create README along with the Dockerfile(s) to highlight or explain some of the design decisions or difficulties that were encountered when created them. 

/cc @kubeflow/wg-notebooks-leads 
Extra ping to @StefanoFioravanzo as requested in https://github.com/kubeflow/kubeflow/issues/5575#issuecomment-797460553.

### CODE REVIEW

File: components/example-notebook-servers/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Kubeflow Example Notebook Servers
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:1]: Remove `Kubeflow` from the title?

File: components/example-notebook-servers/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+**These dockerfiles and the resulting images are provided as examples only. These images do not (yet) undergo automated testing and are only briefly tested manually. As such, these images are not guaranteed to contain no problems. If you do encounter a problem in one of these images, we greatly appreciate contributions to further improve these images.**
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:3]: I would rephrase as follows:

```
**NOTE:** The Notebooks Working Group provides these Dockerfiles as examples only, 
they are not part of our official CI/CD pipeline and are not guaranteed to work in every 
case. As such, we are not providing timely SLA. The Notebooks WG will work on officially 
supporting some of these Dockerfiles in the near future. Contributions and issue reports 
are greatly appreciated.
```

**Review by StefanoFioravanzo — COMMENTED** [review]
Thanks @DavidSpek ! I left a couple of initial comments. Before I continue with the review, could you please refactor the file to stop every line at 80 chars (or more, but no more than 120)? It's hard to apply review comments when every paragraph is a single line. Having shorter lines also helps readability of the source file.

File: components/example-notebook-servers/README.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+**DISCLAIMER:** The Notebooks Working Group provides these Dockerfiles and
+their images as examples only, they are not tested with our CI/CD pipelines
+and are not certified to work in every situation. As such, issues related to
+these images will be dealt with in a best efforts approach. The Notebooks
+Working Group will work on officially supporting some of these Notebook
+Server images in the near future. If you do encounter a problem in one of
+these images, contributions and issue reports are greatly appreciated.
+
+## Introduction
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:11]: I would suggest trying to remove the passive voice in favour for an active voice. This is a general comment for the entire document.

File: components/example-notebook-servers/README.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+This directory contains Dockerfiles that are intended to show users how to
+create images for various types of Notebook Servers that can run on Kubeflow.
+There is a common base images that starts from Ubuntu 20.04, adds some
+dependencies and common utilities from APT, installs `kubectl`, creates the
+user account for the downstream images and adds the
+[S6-overlay](https://github.com/just-containers/s6-overlay) for the init system.
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:18]: Covert this to a bullet list? E.g.:

A common base image provides the following layers:

- Ubuntu 20.04
- `kubectl`
- User account (name?)
- init-system

File: components/example-notebook-servers/README.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+From this base image, images for Jupyter, RStudio and Visual Studio Code
+(through [code-server]([https://github.com/cdr/code-server)) are created.
+From the Jupyter image, images are created that show users how to install
+TensorFlow and PyTorch both with and without support for NVIDIA GPUs.
+In turn, these images are extended with an opinionated set of common tools
+and packages to help new users to quickly get started with Kubeflow.
+
+**Lineage of the provided example Notebook Server Images:**
+
+![Flow Chart of example Notebook Servers](./kubeflow-example-notebook-servers.png)
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:28]: Could you upload an image with a white background? If you use GitHub with the Dark theme it's hard to see the arrow connections

File: components/example-notebook-servers/README.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+a selection of packages specific that will be needed in the Notebook Server.
+Users can install packages in a running Notebook Server using `pip install ...`
+or `conda install ...`. However, packages installed in this manner will not
+persist if the Notebook Server is stopped or the Pod is restarted.
+While it is possible to install Python packages in the user's home directory
+by using `pip install --user ...`, this can cause unexpected behaviour if the
+workspace volume is later used in a different Notebook Server, as the packages
+present in the home directory might be used rather than the ones present in the
+Notebook Server's image. For this reason it is recommended to create custom
+images if persistance of the installed packages is wanted. An example of such
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:48]: Passive to active voice example, I'd change this to: `We recommend .. if you want to ...`

File: components/example-notebook-servers/README.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+a Dockerfile can be seen (POINT TO A DOCKERFILE THAT USES PIP AND ONE THAT USES CONDA).
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:49]: Are you planning on adding a link here?

File: components/example-notebook-servers/README.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+If APT dependencies are required need to be installed, this must be done with by
+creating a custom Notebook Server image as the user in the provided images does
+not have `sudo` privileges. To do so, create a Dockerfile and choose a given
+base image. Next, change to the root user using `USER root` and make any
+changes that require root privileges. Once this is done, switch back to the
+regular user with `USER $NB_UID` and, if needed, install the wanted packages
+using Pip or Conda.
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:56]: Instead of describing this in text, I would just provide a Dockerfile snippet that shows what you are describing

File: components/example-notebook-servers/README.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+The most common customization to Notebook Server images will be to install
+a selection of packages specific that will be needed in the Notebook Server.
+Users can install packages in a running Notebook Server using `pip install ...`
+or `conda install ...`. However, packages installed in this manner will not
+persist if the Notebook Server is stopped or the Pod is restarted.
+While it is possible to install Python packages in the user's home directory
+by using `pip install --user ...`, this can cause unexpected behaviour if the
+workspace volume is later used in a different Notebook Server, as the packages
+present in the home directory might be used rather than the ones present in the
+Notebook Server's image. For this reason it is recommended to create custom
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:47]: I am not sure the discussion about the lifecycle of packages and PVCs is relevant to this document. I would remove it

File: components/example-notebook-servers/README.md [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+regular user with `USER $NB_UID` and, if needed, install the wanted packages
+using Pip or Conda.
+
+### Custom startup scripts and extra services (Advanced)
+
+Some users might require custom scripts to run during the startup of the
+Notebook Server container, or advanced users might want to add additional
+services that run inside the container (for example: an Apache or NGINX web server).
+To make this easy, the [S6-overlay](https://github.com/just-containers/s6-overlay)
+is used in the provided images.
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:64]: Another example of passive to active: `To make this easy, we use the S6-overlay as init-system`

File: components/example-notebook-servers/README.md [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+process running in a container, the
+[S6-overlay](https://github.com/just-containers/s6-overlay) is built to manage
+multiple processes and allows the creator of the image to determine which
+processes should restart and which should cause the container to exit.
+
+#### Custom startup scripts
+
+Scripts that need to run during the startup of the Notebook Server container
+can be placed in the directory `/etc/cont-init.d/`. An example of such a script
+can be found [here](POINT TO RSTUDIO INIT SCRIPT). In this script, the
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:84]: Missing URL

File: components/example-notebook-servers/README.md [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+[JupyterLab](https://github.com/jupyterlab/jupyterlab). For more information on
+creating `run` and `finish` scripts for services please see the
+[S6-overlay documentation](https://github.com/just-containers/s6-overlay#writing-a-service-script).
+
+**Important**
+The example Notebook Server images that are provided here execute the
+[S6-overlay](https://github.com/just-containers/s6-overlay) as user `$NB_USER`
+and not as `root`. This is done so that none of the processes in the container
+are executed as `root` user. As such, any files or scripts related to the
+[S6-overlay](https://github.com/just-containers/s6-overlay) must be owner by
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:111]: `owner` -> `owned`

File: components/example-notebook-servers/README.md [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+user `$NB_USER`. To copy startup or service scripts into the container image
+you can use `--chown` with the `COPY` command in the Dockerfile
+(LINK TO A PLACE WHERE THIS IS USED). As the
```
Comment (StefanoFioravanzo) [inline components/example-notebook-servers/README.md:114]: Missing URL (I know you know - I am just adding a review comment to make sure we don't forget about it)

**Review by StefanoFioravanzo — COMMENTED** [review]
@DavidSpek  I added some more comments on the rest of the doc

File: components/example-notebook-servers/README.md (hunk 3 above)
Comment (DavidSpek) [reply components/example-notebook-servers/README.md:11]: Yeah, I will look into this. I'm used to scientific writing so tend to default to the passive voice.

File: components/example-notebook-servers/README.md (hunk 4 above)
Comment (DavidSpek) [reply components/example-notebook-servers/README.md:18]: I didn't want to go into detail about the images themselves here (leave that to the READMEs for each image), but it does make sense to add the base image specification here. 

The user account is still `jovyan` at the moment (for compatibility reasons), for 1.4 this should probably be changed to `kubeflow`.

File: components/example-notebook-servers/README.md (hunk 7 above)
Comment (DavidSpek) [reply components/example-notebook-servers/README.md:49]: Yep. But I can't yet as it is not merged.

File: components/example-notebook-servers/README.md (hunk 8 above)
Comment (DavidSpek) [reply components/example-notebook-servers/README.md:56]: It wasn't until yesterday evening that I had a Dockerfile that would serve as a good example here that could be linked to (the rstudio-tidyverse image is perfect for this). So I will shorten this a lot and add a reminder to add the link.

File: components/example-notebook-servers/README.md (hunk 9 above)
Comment (DavidSpek) [reply components/example-notebook-servers/README.md:47]: While not directly relevant to this document, this very situation has been the cause of issues in the past so I do think this should be stated somewhere. I agree, it is a bit out of place here. I'll leave it in here for the moment until while I think of what to do with it. If needed, it can quickly be removed before merging. 

I'm thinking a `Best practices and considerations` section might be on option (in this README, as well as on the website, which I'm guessing will be mostly copy/paste from here with some formatting changes).

File: components/example-notebook-servers/README.md (hunk 11 above)
Comment (DavidSpek) [reply components/example-notebook-servers/README.md:84]: Yep, hasn't been merged yet :).

File: components/example-notebook-servers/README.md (hunk 12 above)
Comment (DavidSpek) [reply components/example-notebook-servers/README.md:111]: Good catch!

**Comment by StefanoFioravanzo** [comment]
@DavidSpek I saw you pushed force pushed the branch. Could you please resolve the conversations you addressed or point to the new changes? It would help if you could stack new `review: ...` commits on top of the old ones instead of force pushing, so it's easier to keep track of what changes between revisions. 

What's the state now? Are you done and we should do another pass?

**Comment by DavidSpek** [comment]
@StefanoFioravanzo Sorry for the late response, I missed your comment with all the other things going on. I force pushed simple to keep it up to date with master (so presubmit jobs aren't created for other components). The "review:" commits should be stacking on top of the previous one and the conversations that I've addressed I've marked as resolved. If something isn't correct I'll look into fixing it tomorrow. There are some other changes I want to add to the readme that I'll be looking at tomorrow.

**Comment by DavidSpek** [comment]
@StefanoFioravanzo While the README isn't perfect, as users are starting to install 1.3 and using these images I think it is important to merge this PR quickly. Especially given the disclaimer.

File: components/example-notebook-servers/README.md (hunk 9 above)
Comment (thesuperzapper) [reply components/example-notebook-servers/README.md]: While I think this wording can be simplified, we should definitely discuss the implications of how kubeflow mounts PVCs to the home directory for `pip install`

**Comment by kimwnasptd** [comment]
@StefanoFioravanzo @thesuperzapper @DavidSpek what's left to merge this PR?

I can only see some comments from @StefanoFioravanzo about removing passive voice, which I also agree. Is there something else that needs to be done before merging the PR?

**Comment by DavidSpek** [comment]
@kimwnasptd the links to the dockerfiles now that I think of it. I'll add those and run through it again in a moment.

**Comment by thesuperzapper** [comment]
This was superseded by #5818
/close
